### DISCUSSION TITLE
without database involved, how is job status maintained?

### DISCUSSION THREAD

**Discussion by Xavierxhq** [go-co-op/gocron#770] [Q&A] [answered]
jobs are allowed to be defined as month jobs, considering such a long period, how is the status of jobs managed?
moreover, what if a job failed, will it be re run?

**Comment by JohnRoesler** [comment]
> jobs are allowed to be defined as month jobs, considering such a long period, how is the status of jobs managed?

It's up to your app to decide how to handle maintaining status / state. The design of gocron is pretty resilient such that, if your application were to restart, the job would just be scheduled again for the same next time. For example, you're next run is scheduled for Oct 1, and your application restarts on Sep 29. Gocron would recalculate the jobs next run on start up and would still be Oct 1. The main risk you have to determine how you want to handle is the case where your application goes down on Sep 30 and doesn't come back up until Oct 2. For that case, it would be scheduled for Nov 1 and the Oct 1 run would be missed.

> moreover, what if a job failed, will it be re run?

There are no retries built in. You can determine how you want to handle a "failure" by returning an error from your job and using the [AfterJobRunsWithError](https://pkg.go.dev/github.com/go-co-op/gocron/v2#AfterJobRunsWithError) listener

**Reply by Xavierxhq** [reply]
thx for the answer. 

just about the jobs' state maintaining, I am wondering where the state's data is stored, in memory? or somewhere in the disk?

and one more question just came to my mind, which is that for a cron-job, if the next call arrives while the current one is still in process, will the next call be activated and proceeded? or can I manage in deciding what would happen in such a situation?

thx for your kind reply again~

**Reply by JohnRoesler** [reply, marked as answer]
> just about the jobs' state maintaining, I am wondering where the state's data is stored, in memory? or somewhere in the disk?

in memory

> and one more question just came to my mind, which is that for a cron-job, if the next call arrives while the current one is still in process, will the next call be activated and proceeded? or can I manage in deciding what would happen in such a situation?

Couple options:
- Per job, [WithSingletonMode](https://pkg.go.dev/github.com/go-co-op/gocron/v2#WithSingletonMode)
- Per scheduler, [WithLimitConcurrentJobs](https://pkg.go.dev/github.com/go-co-op/gocron/v2#WithLimitConcurrentJobs)

**Reply by Xavierxhq** [reply]
got it. thx a lot.

**Comment by Xavierxhq** [comment]
Hi there, i am Xavier, if u r seeing this, it means that i have received your letter and i will try to reply sooner. Ur respectfully.
